### ISSUE TITLE
OpenSUSE: missing desktop-file-utils dependency.

### ISSUE THREAD

**Issue by mook-as** [rancher-sandbox/rancher-desktop#1105]
### Rancher Desktop Version

0.7.0beta.1.101.a1caf74

### Rancher Desktop K8s Version

N/A

### What operating system are you using?

Other Linux

### Operating System / Build Version

openSUSE Leap 15.3

### What CPU architecture are you using?

x64

### Actual Behavior

Upon installing `rancher-desktop`:
```
(4/4) Installing: rancher-desktop-0.7.0beta.1.101.a1caf74-86.1.x86_64 …[done]
Additional rpm output:
/var/tmp/rpm-tmp.Of3bhi: line 4: update-desktop-database: command not found
```

### Steps to Reproduce

```bash
zypper addrepo obs://isv:Rancher:dev/rpm/ obs:isv:Rancher:dev
zypper install rancher-desktop
```

### Result

See above.

### Expected Behavior

If we need to run `update-desktop-database`, we should depend on `desktop-file-utils`

### Additional Information

This was on WSL2 (because I'm being silly).

**Comment by mook-as** [comment]
Actually, it looks like we have _very_ minimal information about dependencies:
```sh
# zypper info --requires rancher-desktop
⋮
Requires       : [3]
    openssh-clients
    qemu
    /bin/sh
# # ldd -r /usr/bin/rancher-desktop
        linux-vdso.so.1 (0x00007ffd2edff000)
        libffmpeg.so => not found
        libdl.so.2 => /lib64/libdl.so.2 (0x00007efd5a5fa000)
        libpthread.so.0 => /lib64/libpthread.so.0 (0x00007efd5a3da000)
        librt.so.1 => /lib64/librt.so.1 (0x00007efd5a1d2000)
        libgobject-2.0.so.0 => /usr/lib64/libgobject-2.0.so.0 (0x00007efd59f76000)
        libglib-2.0.so.0 => /usr/lib64/libglib-2.0.so.0 (0x00007efd59c51000)
        libxshmfence.so.1 => /usr/lib64/libxshmfence.so.1 (0x00007efd59a4e000)
        libgio-2.0.so.0 => /usr/lib64/libgio-2.0.so.0 (0x00007efd5967d000)
        libnss3.so => /usr/lib64/libnss3.so (0x00007efd59346000)
        libnssutil3.so => /usr/lib64/libnssutil3.so (0x00007efd59115000)
        libsmime3.so => /usr/lib64/libsmime3.so (0x00007efd58eed000)
        libnspr4.so => /usr/lib64/libnspr4.so (0x00007efd58cae000)
        libatk-1.0.so.0 => /usr/lib64/libatk-1.0.so.0 (0x00007efd58a87000)
        libatk-bridge-2.0.so.0 => /usr/lib64/libatk-bridge-2.0.so.0 (0x00007efd58853000)
        libdbus-1.so.3 => /usr/lib64/libdbus-1.so.3 (0x00007efd58601000)
        libdrm.so.2 => /usr/lib64/libdrm.so.2 (0x00007efd583ef000)
        libgdk_pixbuf-2.0.so.0 => /usr/lib64/libgdk_pixbuf-2.0.so.0 (0x00007efd581ca000)
        libgtk-3.so.0 => /usr/lib64/libgtk-3.so.0 (0x00007efd57832000)
        libgdk-3.so.0 => /usr/lib64/libgdk-3.so.0 (0x00007efd57533000)
        libpango-1.0.so.0 => /usr/lib64/libpango-1.0.so.0 (0x00007efd572e7000)
        libcairo.so.2 => /usr/lib64/libcairo.so.2 (0x00007efd56fb0000)
        libm.so.6 => /lib64/libm.so.6 (0x00007efd56c6f000)
        libX11.so.6 => /usr/lib64/libX11.so.6 (0x00007efd5692e000)
        libXcomposite.so.1 => /usr/lib64/libXcomposite.so.1 (0x00007efd5672b000)
        libXdamage.so.1 => /usr/lib64/libXdamage.so.1 (0x00007efd56528000)
        libXext.so.6 => /usr/lib64/libXext.so.6 (0x00007efd56316000)
        libXfixes.so.3 => /usr/lib64/libXfixes.so.3 (0x00007efd56110000)
        libXrandr.so.2 => /usr/lib64/libXrandr.so.2 (0x00007efd55f05000)
        libexpat.so.1 => /usr/lib64/libexpat.so.1 (0x00007efd55cd3000)
        libxcb.so.1 => /usr/lib64/libxcb.so.1 (0x00007efd55aaa000)
        libxkbcommon.so.0 => /usr/lib64/libxkbcommon.so.0 (0x00007efd5586a000)
        libgbm.so.1 => /usr/lib64/libgbm.so.1 (0x00007efd5565a000)
        libasound.so.2 => /usr/lib64/libasound.so.2 (0x00007efd55362000)
        libcups.so.2 => /usr/lib64/libcups.so.2 (0x00007efd550be000)
        libatspi.so.0 => /usr/lib64/libatspi.so.0 (0x00007efd54e89000)
        libgcc_s.so.1 => /lib64/libgcc_s.so.1 (0x00007efd54c70000)
        libc.so.6 => /lib64/libc.so.6 (0x00007efd5489b000)
        /lib64/ld-linux-x86-64.so.2 (0x00007efd62a2a000)
        libffi.so.7 => /usr/lib64/libffi.so.7 (0x00007efd54691000)
        libpcre.so.1 => /usr/lib64/libpcre.so.1 (0x00007efd54408000)
        libgmodule-2.0.so.0 => /usr/lib64/libgmodule-2.0.so.0 (0x00007efd54204000)
        libz.so.1 => /lib64/libz.so.1 (0x00007efd53fed000)
        libmount.so.1 => /usr/lib64/libmount.so.1 (0x00007efd53d90000)
        libselinux.so.1 => /lib64/libselinux.so.1 (0x00007efd53b67000)
        libresolv.so.2 => /lib64/libresolv.so.2 (0x00007efd53950000)
        libplc4.so => /usr/lib64/libplc4.so (0x00007efd5374b000)
        libplds4.so => /usr/lib64/libplds4.so (0x00007efd53547000)
        libsystemd.so.0 => /usr/lib64/libsystemd.so.0 (0x00007efd53279000)
        libpangocairo-1.0.so.0 => /usr/lib64/libpangocairo-1.0.so.0 (0x00007efd5306a000)
        libXi.so.6 => /usr/lib64/libXi.so.6 (0x00007efd52e59000)
        libcairo-gobject.so.2 => /usr/lib64/libcairo-gobject.so.2 (0x00007efd52c50000)
        libepoxy.so.0 => /usr/lib64/libepoxy.so.0 (0x00007efd52964000)
        libfribidi.so.0 => /usr/lib64/libfribidi.so.0 (0x00007efd52748000)
        libpangoft2-1.0.so.0 => /usr/lib64/libpangoft2-1.0.so.0 (0x00007efd52532000)
        libharfbuzz.so.0 => /usr/lib64/libharfbuzz.so.0 (0x00007efd52229000)
        libfontconfig.so.1 => /usr/lib64/libfontconfig.so.1 (0x00007efd51fe4000)
        libfreetype.so.6 => /usr/lib64/libfreetype.so.6 (0x00007efd51d2a000)
        libXinerama.so.1 => /usr/lib64/libXinerama.so.1 (0x00007efd51b27000)
        libXcursor.so.1 => /usr/lib64/libXcursor.so.1 (0x00007efd5191c000)
        libwayland-cursor.so.0 => /usr/lib64/libwayland-cursor.so.0 (0x00007efd51714000)
        libwayland-egl.so.1 => /usr/lib64/libwayland-egl.so.1 (0x00007efd51512000)
        libwayland-client.so.0 => /usr/lib64/libwayland-client.so.0 (0x00007efd51302000)
        libthai.so.0 => /usr/lib64/libthai.so.0 (0x00007efd510f8000)
        libpixman-1.so.0 => /usr/lib64/libpixman-1.so.0 (0x00007efd50e52000)
        libEGL.so.1 => /usr/lib64/libEGL.so.1 (0x00007efd50c3f000)
        libpng16.so.16 => /usr/lib64/libpng16.so.16 (0x00007efd509fc000)
        libxcb-shm.so.0 => /usr/lib64/libxcb-shm.so.0 (0x00007efd507f8000)
        libxcb-render.so.0 => /usr/lib64/libxcb-render.so.0 (0x00007efd505ea000)
        libXrender.so.1 => /usr/lib64/libXrender.so.1 (0x00007efd503df000)
        libGL.so.1 => /usr/lib64/libGL.so.1 (0x00007efd50154000)
        libXau.so.6 => /usr/lib64/libXau.so.6 (0x00007efd4ff50000)
        libwayland-server.so.0 => /usr/lib64/libwayland-server.so.0 (0x00007efd4fd3c000)
        libgssapi_krb5.so.2 => /usr/lib64/libgssapi_krb5.so.2 (0x00007efd4faf0000)
        libgnutls.so.30 => /usr/lib64/libgnutls.so.30 (0x00007efd4f738000)
        libavahi-common.so.3 => /usr/lib64/libavahi-common.so.3 (0x00007efd4f52b000)
        libavahi-client.so.3 => /usr/lib64/libavahi-client.so.3 (0x00007efd4f31a000)
        libblkid.so.1 => /usr/lib64/libblkid.so.1 (0x00007efd4f0ca000)
        liblzma.so.5 => /usr/lib64/liblzma.so.5 (0x00007efd4ee90000)
        libzstd.so.1 => /usr/lib64/libzstd.so.1 (0x00007efd4ebee000)
        liblz4.so.1 => /usr/lib64/liblz4.so.1 (0x00007efd4e9cd000)
        libcap.so.2 => /usr/lib64/libcap.so.2 (0x00007efd4e7c8000)
        libgcrypt.so.20 => /usr/lib64/libgcrypt.so.20 (0x00007efd4e4aa000)
        libgraphite2.so.3 => /usr/lib64/libgraphite2.so.3 (0x00007efd4e284000)
        libbz2.so.1 => /usr/lib64/libbz2.so.1 (0x00007efd4e067000)
        libdatrie.so.1 => /usr/lib64/libdatrie.so.1 (0x00007efd4de60000)
        libGLdispatch.so.0 => /usr/lib64/libGLdispatch.so.0 (0x00007efd4dbaa000)
        libGLX.so.0 => /usr/lib64/libGLX.so.0 (0x00007efd4d979000)
        libkrb5.so.3 => /usr/lib64/libkrb5.so.3 (0x00007efd4d69d000)
        libk5crypto.so.3 => /usr/lib64/libk5crypto.so.3 (0x00007efd4d46b000)
        libcom_err.so.2 => /lib64/libcom_err.so.2 (0x00007efd4d267000)
        libkrb5support.so.0 => /usr/lib64/libkrb5support.so.0 (0x00007efd4d05a000)
        libp11-kit.so.0 => /usr/lib64/libp11-kit.so.0 (0x00007efd4cdf7000)
        libtasn1.so.6 => /usr/lib64/libtasn1.so.6 (0x00007efd4cbe4000)
        libnettle.so.6 => /usr/lib64/libnettle.so.6 (0x00007efd4c9ab000)
        libhogweed.so.4 => /usr/lib64/libhogweed.so.4 (0x00007efd4c773000)
        libgmp.so.10 => /usr/lib64/libgmp.so.10 (0x00007efd4c4dd000)
        libidn2.so.0 => /usr/lib64/libidn2.so.0 (0x00007efd4c2c0000)
        libunistring.so.2 => /usr/lib64/libunistring.so.2 (0x00007efd4bf3d000)
        libgpg-error.so.0 => /usr/lib64/libgpg-error.so.0 (0x00007efd4bd1d000)
        libkeyutils.so.1 => /usr/lib64/libkeyutils.so.1 (0x00007efd4bb19000)
```
(This was after I installed `chromium` because I got tired of chasing down deps.)

**Comment by adamkpickering** [comment]
```
(4/4) Installing: rancher-desktop-0.7.0beta.1.101.a1caf74-86.1.x86_64 …[done]
Additional rpm output:
/var/tmp/rpm-tmp.Of3bhi: line 4: update-desktop-database: command not found
```
I can't reproduce this. @mook-as is this still an issue? I'm going to fix the dependencies part but I'm not sure I can fix this part.

**Comment by mook-as** [comment]
@adamkpickering It's best to try to reproduce this in a docker container; if you're just doing this on your desktop, you'd already have the things (you probably have a web browser of some sort, which probably pulled in the same dependencies).

**Comment by adamkpickering** [comment]
I have been able to get deb and rpm packages to build with a set of dependencies that more accurately represents the reality. However, I have run into a problem that I'm having a really hard time solving.

RD is built as a dynamically linked executable. As shown above by Mark, it links to a bunch of different libraries. According to `ldd`, each one of these libraries is specified with a major version (i.e. a "soname") in the binary. This is also how it appears to be when looking at the output produced by running RD with `LD_DEBUG=versions` prepended.

The strange thing is that the same package (and thus binary) can be installed on distros that have different major versions of certain libraries available. For example, on Ubuntu 18.04 it links against `libhogweed.so.4`, whereas on Mint 20.3 it links against `libhogweed.so.5`. This is very confusing to me, since I was under the impression that ELF binaries referenced sonames, i.e. the name of the library including the major version. If this is true, running RD on Ubuntu 18.04 or Mint 20.3 would not work. The only way this could work is for the RD ELF binary to reference the library name without any version. But if that is the case, why does `ldd` say that the binary links against `libhogweed.so.4` and `libhogweed.so.5` rather than just `libhogweed.so`?

This is a problem: when listing dependencies in the deb/rpm, those dependencies are packaged by major version. So we can't just list `libhogweed` as a dependency - we have to require `libhogweed4`, or `libhogweed5`. And `libhogweed5` is not available on Ubuntu 18.04, for example.

Originally I compiled a list of required packages based on the output of `ldd` run on Mint 20.3. When the deb built with these requirements is installed on Ubuntu 18.04, there are three libraries that are unavailable because only an earlier major version is available. This is only between two distros - what about others? What about new versions of distros that people will want to use RD on in the future?

I can think of three solutions:
1. Don't track dependencies. This is more of a continuum - we could track some, and only remove the ones that are a problem. Or we could track none except for the ones that cause a problem when not included.
2. Build a different package for every version of every OS. This is obviously not tenable.
3. Get the user to install repositories that will provide the needed dependencies on older distros. I don't think this is tenable either.

**Comment by adamkpickering** [comment]
It turns out that `ldd` lists not only the direct dependencies of a file, but all of the dependencies of dependencies and so on. It is for getting an idea of all the things that a dynamically linked binary needs to run, not just the so's it directly requires. Something like `objdump` or `readelf` is what one should use in this scenario.
